lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -0,0 +1,65 @@
+// lib/Tranforms/IPO/StructFieldCacheAnalysis.cpp - Performs Cache-Aware Structure Analysis- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===-----------------------------------------------------------------------------------------------===//
+//
+// This pass performs analysis on cache-aware structure field accesses based on the following paper
+// and reports recommendations on changes to make on the source code to improve performance.
+//  [1] M. Hagog, C. Tice “Cache Aware Data Layout Reorganization Optimization in GCC”, Proceedings
+//      of the GCC Developers’ Summit,  Ottawa, 2005.
+//
+//===-----------------------------------------------------------------------------------------------===//
+
+#include "llvm/Transforms/IPO/StructFieldCacheAnalysis.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/IPO.h"
+#include "llvm/Analysis/BlockFrequencyInfo.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "struct-field-cache-analysis"
+
+namespace{
+
+class StructFieldCacheAnalysisPass : public ModulePass {
+ public:
+  static char ID;
+  StructFieldCacheAnalysisPass() : ModulePass(ID) {
+    initializeStructFieldCacheAnalysisPassPass(*PassRegistry::getPassRegistry());
+  }
+ private:
+  bool runOnModule(Module &M) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<BlockFrequencyInfoWrapperPass>();
+  }
+
+};
+}
+
+char StructFieldCacheAnalysisPass::ID = 0;
+INITIALIZE_PASS_BEGIN(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
```

Comment:
No need for begin and end if there are no dependent passes (INITIALIZE_PASS will be fine). But of course you want to depend on BlockFrequencyInfo at the least, so this is fine.